File: project_euler/Problem095/Problem095.java
package project_euler;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Problem095 {
    public void solution() {
        int target = 1000000;
        long[] divisors = new long[target];
        Arrays.fill(divisors, -1L);
        for (int i = 2; i <= target; ++i) {
            getDivisorSummation(i, divisors, target);
        }
        for (int i = 1; i < target; ++i) {
            divisors[i] -= i + 1;
        }

        int result = 0;
        int pivot = 0;
        Set<Integer> chains = new HashSet<>();
        for (int i = 2; i <= target; ++i) {
            int count = getChain(i, divisors, chains, target);
            if (pivot < count) {
                pivot = count;
                result = target;
                for (int value : chains) {
                    if (result > value) {
                        result = value;
                    }
                }
            }
        }
        System.out.println(result);
    }

    private void getDivisorSummation(int value, long[] divisors, int length) {
        if (-1 == divisors[value - 1]) {
            long index = value;
            int count = 1;
            while (index <= length) {
                divisors[(int) index - 1] = ((long) Math.pow(value, count + 1) - 1) / (value - 1);
                index *= value;
                ++count;
            }

            for (int i = Math.min(value - 1, length / value); i >= 2; --i) {
                divisors[value * i - 1] = divisors[value - 1] * divisors[i - 1];
            }
        } else {
            for (int i = Math.min(value - 1, length / value); i >= 2; --i) {
                if (isRelativelyPrime(value, i) && -1 == divisors[value * i - 1]) {
                    divisors[value * i - 1] = divisors[value - 1] * divisors[i - 1];
                }
            }
        }
    }

    private int getChain(long value, long[] divisors, Set<Integer> chains, int target) {
        chains.clear();
        chains.add((int) value);

        long pivot = value;
        int count = 0;
        while (true) {
            value = divisors[(int) value - 1];
            if (pivot == value) {
                break;
            }
            if (1 == value || value >= target || chains.contains((int) value)) {
                return 0;
            }

            chains.add((int) value);
            ++count;
        }
        return count;
    }

    private boolean isRelativelyPrime(int i, int j) {
        while (0 != i % j) {
            int auxiliary = i % j;
            i = j;
            j = auxiliary;
        }
        return 1 == j;
    }
}

File: project_euler/Problem095/Problem095.py
from typing import List

class Problem095:
    def solution(self) -> None:
        target = 1000000
        divisors = [-1 for _ in range(target)]
        for i in range(2, target + 1):
            self.__get_divisor_summation(i, divisors, target)
        for i in range(1, target):
            divisors[i] -= i + 1

        result, pivot, chains = 0, 0, set()
        for i in range(2, target + 1):
            count = self.__get_chain(i, divisors, chains, target)
            if pivot < count:
                pivot, result = count, target
                for value in chains:
                    if result > value:
                        result = value
        print(result)

    def __get_divisor_summation(self, value: int, divisors: List[int], length: int) -> None:
        if -1 == divisors[value - 1]:
            index, count = value, 1
            while index <= length:
                divisors[index - 1] = (value ** (count + 1) - 1) // (value - 1)
                index *= value
                count += 1

            for i in range(min(value - 1, length // value), 1, -1):
                divisors[value * i - 1] = divisors[value - 1] * divisors[i - 1]
        else:
            for i in range(min(value - 1, length // value), 1, -1):
                if self.__is_relatively_prime(value, i) and -1 == divisors[value * i - 1]:
                    divisors[value * i - 1] = divisors[value - 1] * divisors[i - 1]

    def __get_chain(self, value: int, divisors: List[int], chains: set[int], target: int) -> int:
        chains.clear()
        chains.add(value)

        pivot, count = value, 0
        while True:
            value = divisors[value - 1]
            if pivot == value:
                break
            if 1 == value or value >= target or value in chains:
                return 0

            chains.add(value)
            count += 1
        return count

    def __is_relatively_prime(self, i: int, j: int) -> bool:
        while i % j:
            i, j = j, i % j
        return 1 == j
